Minor refactor

meyer opened the pull request:
A few little tweaks and a few major ones. My objective with this PR is not to get it merged as-is, but rather to start a discussion. The changes I’ve made have been in the interest of simplicity, “““elegance”””, and reducing the need for frequent updates to rwb.
## Notable Changes
- **react-router has been removed from rwb** for two reasons:
  1. react-router is complete overkill for a single component setup
  2. react-router has gone through so many breaking changes that bundling it with something foundational like rwb seems a not-so-great idea.
- **Static rendering has been offloaded to the project folder.** Turns out, rwb is a really powerful static site generator. The static output was decent, but I wanted control over more than just component output. Letting the project handle static site generation means that you can use neat components like like [react-helmet](https://github.com/nfl/react-helmet) or [react-document-title](https://github.com/gaearon/react-document-title). This addresses concerns raised in PR #14 and #15.
  
  Details:
  - The static renderer has access to real-deal `fs`.
  - The DOM node and ID are configurable. You can specify `span#wow` in your project’s package.json and it’ll mount `rwb.main` to `<span id="wow"></span>`. Neat!
  - Generated asset filenames use `[hash]`. Kind of hacked in, but it’s decent.
  - Paths and asset filenames are available to the static renderer via the `RWB` object. No more hard-coded `bundle.js`!
- **Loaders and react/react-dom can be updated by the user without any changes to rwb.** No more `react` alias in the webpack config! In hindsight it was a simple change, but it took me ages to get there. I am not a clever man.
## Minor Changes
- **The example has undergone a minor overhaul.** Mainly to show that you can write ES6 javascript and that you can and should use CSS modules. The included `Rakefile` has a `rebuild_example` task that makes rebuilding the example directory a one-line thing. Could’ve written it as a shell script, I guess, but rake was the most cross-platform way of parsing and writing JSON that I know of.
- **`NODE_ENV` is configurable.** It should probably be set with a boolean `RWB_PROD` option (would you ever need anything other than `production` and `development`?) but whateverrrr.
## A few things I’m still thinking about
- Can this be updated to use latest node features (`let`/`const`, `=>`, etc.), or does it need to stay super backwards compatible?
- It’d be super cool if some of the webpack config (maybe just loaders?) was offloaded to the project folder as well. Right now, I can use any loader I want, but I have to prefix whatever I’m requiring with the loader string, which gets a bit messy. The main issue I see right now is that there’s a slight difference between `serve` and `static` loaders due to the hot reloading biz.

Hope this all makes sense, and my apologies for the novella.

meyer committed:
package updates, example folder updates

meyer committed:
WIP: server rendering all routes, entrypoint moved to project dir

Coming soon: server rendering handled by the individual project instead
of rwb

meyer committed:
WIP: moved static rendering/index to project dir

this will be gross until I can figure out the issue with `target: 'node'` mentioned in Workflows.js:251

meyer committed:
WIP: static generator now fully moved over to project dir

meyer committed:
Example updates, static gen improvements

Chunked assets are now available to the static page generator as
`__rwbAssets`.

The locations of the project folder and static folder are available as
`__rwbProjectRoot` and `__rwbStaticRoot`.

meyer committed:
Moved entrypoint back to rwb, improved static rendering

also, check out that Rakefile. daaaaaaaamn

meyer committed:
fixed silly `rwb serve` bug

meyer committed:
Minor loader tweaks

meyer committed:
Resolve root magic, package updates

required modules can now be overridden by the individual project! no
more of that react alias bullshit.

meyer committed:
Added svgo loader with some default options

meyer committed:
Updated loaders, added sourcemaps to `css-loader`

meyer committed:
Made asset path and NODE_ENV configurable

meyer committed:
Fixed hot reloading bug in example

meyer committed:
Fixed double slash issue in example

meyer committed:
Restored accidental deletion from 8d9727c

oops.

meyer committed:
Removed lint, fixed charset issue in example

meyer committed:
Added RWB_DISABLE_CACHEBUSTER option

meyer committed:
Fixed absolute path issue

Aourin commented:
I think one of biggest things are the ability to configure loaders. However, that might break the original vision of being able to standardize the comps. The problem with standardization of stuff like this is that there is no great way to do it because everyone will want to use loaders however they want so if someone is using a generic repo created by rwb, they'll have to make sure to use the same loaders and what not.

meyer committed:
Update package.json + example project + Rakefile

meyer committed:
Merge remote-tracking branch 'upstream/master' into opinions

# Conflicts:
#	.gitignore
#	example/MyComponent.css
#	example/MyComponent.js
#	example/package.json
#	example/routes.js
#	lib/RWB.js
#	lib/Workflows.js
#	lib/entrypoint.js
#	lib/template/routes.js
#	package.json

meyer committed:
react-hot-loader 3.0.0b2, ditch ReactElement idea

meyer committed:
Cleanup

meyer committed:
Update to ES6, add eslintrc

meyer committed:
Don’t overwrite scripts.start in package.json

meyer committed:
Remove default export workaround

No longer necessary thanks to babel-plugin-add-module-exports :sunglasses:

meyer committed:
ES6 template strings baybeeee

meyer committed:
Update deps + cleanup, Node 5 compatibility fix

meyer committed:
Remove browser open feature

meyer committed:
Remove JSX from entrypoint, clean up example a bit

meyer committed:
Validate arguments with yargs

meyer committed:
Split Workflows.js into separate files

meyer committed:
Move webpack-specific stuff to subfolder

meyer committed:
StandardWebpack.validate → validateConfig

meyer committed:
Fix bug in generateConfig method

meyer committed:
Improve webpack config

- whitelisting with `include` instead of blacklisting
- camelcase filenames to match exported functions
- muffle yargs help message on error

meyer committed:
Fix overzealous `showHelpOnFail`, tweak options
